openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,559 @@
+			return true
+		}
+	}
+
+	return false
+}
+
+func getWorkloadType(annotations map[string]string) (string, error) {
+	var workloadAnnotationsKeys []string
+	for k := range annotations {
+		if strings.HasPrefix(k, podWorkloadTargetAnnotationPrefix) {
+			workloadAnnotationsKeys = append(workloadAnnotationsKeys, k)
+		}
+	}
+
+	// no workload annotation is specified under the pod
+	if len(workloadAnnotationsKeys) == 0 {
+		return "", nil
+	}
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotationsKeys) > 1 {
+		return "", fmt.Errorf("the pod can not have more than one workload annotations")
+	}
+
+	workloadType := strings.TrimPrefix(workloadAnnotationsKeys[0], podWorkloadTargetAnnotationPrefix)
+	if len(workloadType) == 0 {
+		return "", fmt.Errorf("the workload annotation key should have format %s<workload_type>, when <workload_type> is non empty string", podWorkloadTargetAnnotationPrefix)
+	}
+
+	return workloadType, nil
+}
+
+func getWorkloadAnnotationEffect(workloadAnnotationKey string) (string, error) {
+	managementAnnotationValue := map[string]string{}
+	if err := json.Unmarshal([]byte(workloadAnnotationKey), &managementAnnotationValue); err != nil {
+		return "", fmt.Errorf("failed to parse %q annotation value: %v", workloadAnnotationKey, err)
+	}
+
+	if len(managementAnnotationValue) > 1 {
+		return "", fmt.Errorf("the workload annotation value %q has more than one key", managementAnnotationValue)
+	}
+
+	effect, ok := managementAnnotationValue[podWorkloadAnnotationEffect]
+	if !ok {
+		return "", fmt.Errorf("the workload annotation value %q does not have %q key", managementAnnotationValue, podWorkloadAnnotationEffect)
+	}
+	return effect, nil
+}
+
+func stripResourcesAnnotations(annotations map[string]string) {
+	for k := range annotations {
+		if strings.HasPrefix(k, containerResourcesAnnotationPrefix) {
+			delete(annotations, k)
+		}
+	}
+}
+
+func stripWorkloadAnnotations(annotations map[string]string) {
+	for k := range annotations {
+		if strings.HasPrefix(k, podWorkloadTargetAnnotationPrefix) {
+			delete(annotations, k)
+		}
+	}
+}
+
+func addWorkloadAnnotations(annotations map[string]string, workloadType string) {
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadTargetAnnotationPrefix, workloadType)
+	annotations[workloadAnnotation] = fmt.Sprintf(`{"%s":"%s"}`, podWorkloadAnnotationEffect, workloadEffectPreferredDuringScheduling)
+}
+
+func (a *managementCPUsOverride) Validate(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
```

Comment:
Add a skip here if the pod is a static pod.

Reply:
see my comment above.

Reply:
done